File: day22/src/main.py
from collections import defaultdict
from itertools import pairwise
from pathlib import Path

MASK_24BIT = 0xFFFFFF
GENS = 2000
SEQ_LEN = 4


def part1():
    data = Path('my_input.txt').read_text().strip().splitlines()
    secrets = map(int, data)

    secret_sum = 0
    for secret in secrets:
        for _ in range(GENS):
            secret = ((secret << 6) ^ secret) & MASK_24BIT
            secret = ((secret >> 5) ^ secret) & MASK_24BIT
            secret = ((secret << 11) ^ secret) & MASK_24BIT

        secret_sum += secret

    print("Answer:", secret_sum)


def part2():
    data = Path('my_input.txt').read_text().strip().splitlines()
    secrets = map(int, data)

    amounts = defaultdict(int)
    for secret in secrets:
        digits = []
        for _ in range(GENS):
            secret = ((secret << 6) ^ secret) & MASK_24BIT
            secret = ((secret >> 5) ^ secret) & MASK_24BIT
            secret = ((secret << 11) ^ secret) & MASK_24BIT
            digits.append(secret % 10)

        deltas = [b - a for a, b in pairwise(digits)]
        deltas_windows = (tuple(deltas[i: i + SEQ_LEN]) for i in range(GENS - SEQ_LEN))
        seen_sequences = set()
        for sequence, price in zip(deltas_windows, digits[SEQ_LEN:]):
            if sequence not in seen_sequences:
                seen_sequences.add(sequence)
                amounts[sequence] += price

    print("Answer:", max(amounts.values()))


if __name__ == '__main__':
    part1()
    part2()

File: day22/src/main.rs
use itertools::Itertools;
use std::collections::{HashMap, HashSet};
use std::fs;

const MASK_24BIT: i64 = 0xFFFFFF;
const GENS: usize = 2000;
const SEQ_LEN: usize = 4;

fn part1() {
    let data = fs::read_to_string("my_input.txt").expect("Failed to read file");
    let secrets: Vec<i64> = data.lines().map(|line| line.parse().unwrap()).collect();

    let mut secret_sum = 0;
    for mut secret in secrets {
        for _ in 0..GENS {
            secret = ((secret << 6) ^ secret) & MASK_24BIT;
            secret = ((secret >> 5) ^ secret) & MASK_24BIT;
            secret = ((secret << 11) ^ secret) & MASK_24BIT;
        }
        secret_sum += secret;
    }

    println!("Answer: {}", secret_sum);
}

fn part2() {
    let data = fs::read_to_string("my_input.txt").expect("Failed to read file");
    let secrets: Vec<i64> = data.lines().map(|line| line.parse().unwrap()).collect();

    let mut amounts: HashMap<Vec<i64>, i64> = HashMap::new();
    for mut secret in secrets {
        let mut digits: Vec<i64> = vec![];
        for _ in 0..GENS {
            secret = ((secret << 6) ^ secret) & MASK_24BIT;
            secret = ((secret >> 5) ^ secret) & MASK_24BIT;
            secret = ((secret << 11) ^ secret) & MASK_24BIT;
            digits.push(secret % 10);
        }
        let deltas: Vec<i64> = digits.iter().tuple_windows().map(|(a, b)| b - a).collect();
        let mut seen_sequences = HashSet::new();
        for (key, &price) in deltas.windows(SEQ_LEN).zip(digits.iter().skip(SEQ_LEN)) {
            if seen_sequences.insert(key) {
                *amounts.entry(key.to_vec()).or_insert(0) += price;
            }
        }
    }

    println!("Answer: {}", amounts.values().max().unwrap());
}

fn main() {
    part1();
    part2();
}
